library/core/src/pin.rs

```diff
@@ -110,79 +221,221 @@
+//!
+//! <!-- NOTE from anp: "until further notice" is awkward phrasing but i assume you've thought a bit
+//!      about it. consider "until destruction allows fixing up pointers" or similar? -->
 //!
-//! In an intrusive doubly-linked list, the collection does not actually allocate
-//! the memory for the elements itself. Allocation is controlled by the clients,
-//! and elements can live on a stack frame that lives shorter than the collection does.
+//! 1.  *Address Stability.* If [`unsafe`] code witnesses any [`p: Pin<P>`][Pin] at any time then
+//!     it may assume that `p.as_ref().get_ref() as *const _` will remain valid, pointing to the
+//!     same object until further notice.
+//! 2.  *Notice of Destruction.* If `x: T` was ever reachable through any [`Pin<P>`] type, its
+//!     destructor must be run (until it either returns or panics) before `x`'s storage can be
+//!     overwritten. The "until further notice" in (1) includes this mandatory destruction. This is
+//!     often called the "[`Drop`] guarantee".
 //!
-//! To make this work, every element has pointers to its predecessor and successor in
-//! the list. Elements can only be added when they are pinned, because moving the elements
-//! around would invalidate the pointers. Moreover, the [`Drop`][Drop] implementation of a linked
-//! list element will patch the pointers of its predecessor and successor to remove itself
-//! from the list.
+//! ## Address Stability
+//! [address-stability]: #address-stability
 //!
-//! Crucially, we have to be able to rely on [`drop`] being called. If an element
-//! could be deallocated or otherwise invalidated without calling [`drop`], the pointers into it
-//! from its neighboring elements would become invalid, which would break the data structure.
+//! The precise meaning of "address stability" is subtle, because "the same object" is not well-defined.
+//! It is easiest to reason about it in terms of *visibility of mutations*. If [`unsafe`] code mutates
+//! through a [`Pin<P>`], all code that stashed a raw pointer into it will see the mutation. In other
+//! words, [`unsafe`] code can rely on the same value in memory being updated by all uses of a particular
+//! [`Pin<P>`], not to mention that those stashed raw pointers remain valid in the first place.
 //!
-//! Therefore, pinning also comes with a [`drop`]-related guarantee.
+//! When a [`List`] stores a [`Node`], it needs to assume that appending a second node will mutate the
+//! first node, so that later, when the first node is removed, it knows that its predecessor is the
+//! second node.
 //!
-//! # `Drop` guarantee
+//! When writing generic code, it's not possible to know what [`unsafe`] code has recorded about the
+//! pointee's address, so it must be very careful to observe this invariant. Thankfully, most of this
+//! is already enforced by the [`Pin<P>`] API, so only [`unsafe`] code needs to worry about this.
```

Comment:
Maybe a better way to phrase this is that what pinning _actually_ does is simply prevent safe code from performing any mutations whatsoever to the memory location without running that memory location's destructor first. The mutations that are prevented are typically thought of as types of moves, which is why we use "prevents moves" as the terminology for this, but it is not inherently incorrect for unsafe code to allow you to `swap` two pinned values, assuming that the unsafe block doesn't violate the invariants of some other unsafe blocks.

Reply:
I'm not sure I agree. The intent of this section is to be extremely pedantic about what guarantees are made to unsafe code, rather than the emergent properties of the behavior you describe. This topic is complex enough as it is, and I'd rather not have people have to deduce the "I can record the address and no one will mess with it" lemma from that property.